### PR TITLE
HPCC-19721 Split canMatch() from transform() for disk and index read

### PR DESCRIPTION
**Pull request by ghalliday**
## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

### CODE REVIEW

**Comment by ghalliday** [comment]
@shamser @richardkchapman please review.  2 other commits are from PRs that are required.

File: ecl/hql/hqlpmap.cpp [hunk 1]
```diff
@@ -990,7 +990,13 @@ static bool isTrivialTransform(IHqlExpression * expr, IHqlExpression * selector,
                         return false;
                     IHqlExpression * rightField = rhs->queryChild(1);
                     if (leftField != rightField)
-                        return false;
+                    {
+                        if (!allowDeserialize)
+                            return false;
+                        //MORE
+                        if (true)
```
Comment (shamser) [inline ecl/hql/hqlpmap.cpp]: @ghalliday Don't understand what this code is trying to do here.  Will new code be replacing this very soon?

**Review by shamser — CHANGES_REQUESTED** [review]
@ghalliday Couple of questions.  Also, there is a conflict now (probably, related to the merging of the dependant branch into master).

File: ecl/hqlcpp/hqlsource.cpp [hunk 2]
```diff
@@ -2613,6 +2708,35 @@ void SourceBuilder::buildNormalizeHelpers(IHqlExpression * expr)
...
+{
+    if (extractCanMatch)
+    {
+        MemberFunction func(translator, instance->startctx);
+        func.start("virtual bool canMatch(const void * _left) override");
+        func.ctx.addQuotedLiteral("unsigned char * left = (unsigned char *)_left;");
+        if (newDiskReadMapping)
+            translator.bindTableCursor(func.ctx, tableExpr, "left");
+        else
+            translator.bindTableCursor(func.ctx, tableExpr, "left");
```
[the remark this answers is not in the document]
Comment (ghalliday) [reply ecl/hqlcpp/hqlsource.cpp]: Yes - this is changed in the next PR.

File: ecl/hql/hqlpmap.cpp (hunk 1 above)
Comment (ghalliday) [reply ecl/hql/hqlpmap.cpp]: In a subsequent PR it will allow serialized->deserialized if the flag is passed correctly.

**Comment by richardkchapman** [comment]
@shamser Please review

**Review by shamser — APPROVED** [review]
@richardkchapman Code reviewed. Looks good.

**Pull request merged**